torch/jit/__init__.py

```diff
@@ -252,6 +253,142 @@ def _verify_equal(xs, ys):
+                        tensor_compare_errors += 'Node:\n' + indent(str(n_mod)) + '\n'
+                        compare_stack = n_mod.getSourceLocation()
+                        if compare_stack:
+                            tensor_compare_errors += 'Source Location:\n' + indent(compare_stack) + '\n'
+                        tensor_compare_errors += 'Comparison exception: ' + indent(str(e))
+
+                        break  # For now, only print the first diverging pair
+
+            nondeterministic_ops_warning = None
+            nondeterm_ops = [op for op in module.graph.nodes() if op.isNondeterministic()]
+            if len(nondeterm_ops) > 0:
+                nondeterministic_ops_warning = "Trace had nondeterministic nodes. Nodes:\n"
+                for op in nondeterm_ops:
+                    nondeterministic_ops_warning += indent(str(op))
+                nondeterministic_ops_warning += "\nThis may cause errors in trace checking. To disable trace checking,"\
+                                                " pass disable_checks=True to torch.jit.trace()"
+
+            return graph_diff_errors, tensor_compare_errors, nondeterministic_ops_warning
+
+        def wrap_retval(x):
+            return x if isinstance(x, tuple) else (x,)
+
+        def run_mod_and_filter_tensor_outputs(mod, inputs):
+            outs = wrap_retval(mod(*_clone_inputs(inputs)))
+            outs = [out for out in outs if isinstance(out, torch.Tensor)]
+            return outs
+
+        try:
+            traced_outs = run_mod_and_filter_tensor_outputs(module, inputs)
+        except Exception as e:
+            msg = 'Encountered an exception while running trace with check inputs.\nException:\n' + indent(str(e))
+            raise TracingCheckError(*graph_diagnostic_info(), extra_msg=msg)
+
+        try:
+            fn_outs = run_mod_and_filter_tensor_outputs(func, inputs)
+            for orig, check in zip(traced_outs, fn_outs):
+                torch.testing.assert_allclose(orig.double(), check.double(), rtol=check_tolerance,
+                                              atol=torch.testing._get_default_tolerance(orig, check)[1])
+        except (RuntimeError, AssertionError) as e:
+            # TODO: interpose on tracing the function again and check for
+            # divergence? then we can point to where in the source code
+            # we start diverging in python v.s. the trace
+            msg = 'ERROR: Traced function outputs do not match the Python function outputs.\nException: ' + str(e)
+            raise TracingCheckError(*graph_diagnostic_info(),
+                                    extra_msg=msg)
+
+        try:
+            check_outs = run_mod_and_filter_tensor_outputs(check_mod, inputs)
+        except Exception as e:
+            msg = 'Encountered an exception while running checking trace with check inputs.\nException:\n' \
+                + indent(str(e))
+            raise TracingCheckError(*graph_diagnostic_info(), extra_msg=msg)
```

Comment:
Those 3 try block are still repeated (once you move the `allclose` assertion out of the second one).